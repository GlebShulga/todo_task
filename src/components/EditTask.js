import React, { useState } from "react";
import "../assets/scss/EditTask.scss";

const EditTask = ({
  task,
  patchTask,
  fetchTaskList,
  setIsNewTaskCreated,
  isNewTaskCreated,
  setEditingTaskMode,
}) => {
  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [lengthError, setLengthError] = useState(false);

  console.log(title,'title')

  const onChangeTitle = (e) => {
    setTitle(e.target.value);
    setLengthError(false);
  };

  const onChangeDescription = (e) => {
    setDescription(e.target.value);
  };

  return (
    <div className="EditTask">
      <div className="">
        <div className="EditTask-buttons">
          <button
            type="button"
            className="CreateTask-form_button"
            onClick={async (taskId, title, description, status) => {
              if (title?.length <= 20 && title?.length >= 3) {
                await patchTask(taskId, status, title, description);
                await fetchTaskList();
                setEditingTaskMode(false);
              } else {
                setLengthError(true);
              }
            }}
          >
            Save changes
          </button>
          <button
            type="button"
            className="EditTask-buttons_close"
            onClick={() => {
              setEditingTaskMode(false);
            }}
          >
            Close
          </button>
        </div>
        <div className="EditTask-title">
          <input
            className="EditTask-title_input"
            type="text"
            value={title}
            onChange={onChangeTitle}
            placeholder={task.title}
          />
          {lengthError && (
            <div className="EditTask-Error">
              The task length must not be shorter than 3 characters and must not
              exceed 20 characters.
            </div>
          )}
        </div>
        <div>
          <input
            type="checkbox"
            id="status"
            name="status"
            className=""
            // checked={data.status === done ?? false}
            // value={data.read_only ?? false}
            // onChange={""}
          />
          <label for="status">Done</label>
        </div>
        <input
          className="EditTask-description_input"
          type="text"
          value={description}
          onChange={onChangeDescription}
          placeholder={
            task.description === ""
              ? "Write description of your task"
              : task.description
          }
        />
      </div>
    </div>
  );
};

export default EditTask;
