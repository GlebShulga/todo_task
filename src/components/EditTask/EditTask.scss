.EditTask {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.EditTask-buttons {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
}

.EditTask-buttons_close {
  padding: 0.5rem 1rem;
  margin: 0 0.5rem;
  background-color: grey;
  color: white;
  font-weight: 700;
  &:hover {
    background-color: lightslategray;
  }
}

.EditTask-title {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
}

.EditTask-title_input {
  padding: 0.5rem 1rem;
  border: lightgrey;
  background-color: rgb(235, 235, 235);
  outline: none;

  &:focus {
    background-color: darkgrey;
  }
}

.EditTask-description_input {
  margin-top: 1rem;
  border: solid thin;
  border-color: lightskyblue;
  background-color: rgb(235, 235, 235);
  width: 90%;
  height: 100%;
  font-size:large;
  padding: 0.5rem 0.5rem;
  &:focus {
    background-color: white;
  }
  &::placeholder {
    font-size: 1rem;
position: absolute;
top: 0.5rem; left: 0.5rem;
  }
}

.EditTask-Error {
  color: red;
  padding-top: 0.3rem;
}
