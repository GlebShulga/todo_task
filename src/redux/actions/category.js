/* eslint-disable no-loop-func */
import axios from "axios";
import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
} from "../types/category";

export function fetchCategoryList() {
  return (dispatch, getState) => {
    const store = getState();
    const url = store.router.location.pathname?.substring(1);
    const complexUrl = url?.match("[^/](.*?)(?=/)");
    const categoryUrl = complexUrl !== null ? complexUrl[0] : url;
    return axios
      .get("/api/v1/category")
      .then(({ data }) => {
        let current = data.find((cat) => cat.categoryTitle === categoryUrl);
        let updatedCategoryList = data;
        while (current?.parentCategoryId) {
          updatedCategoryList = updatedCategoryList.map((cat) =>
            cat.categoryId === current.parentCategoryId
              ? { ...cat, isExpanded: true }
              : cat
          );
          current = updatedCategoryList.find(
            (cat) => cat.categoryId === current.parentCategoryId
          );
        }

        const rootList = updatedCategoryList.filter(
          (el) => el.parentCategoryId === null
        );
        const titleList = data.map((category) => category.categoryTitle);
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
        dispatch({
          type: GET_CATEGORIES_LIST,
          categoryList: updatedCategoryList,
        });
        dispatch({ type: GET_CATEGORIES_TITLE_LIST, titleList });
      })
      .catch((err) => console.log(err));
  };
}

export function addCategory(categoryTitle, parentCategoryId, lvl) {
  return (dispatch) => {
    axios({
      method: "post",
      url: "/api/v1/category/:categoryTitle",
      data: {
        categoryTitle,
        parentCategoryId,
        lvl,
      },
    })
      .then(({ data: categoryList }) => {
        const rootList = categoryList?.filter(
          (el) => el.parentCategoryId === null
        );
        dispatch({ type: ADD_CATEGORY, categoryList });
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
      })
      .catch((err) => console.log(err));
  };
}

export function renameCategory(categoryId, categoryTitle) {
  return (dispatch) => {
    axios({
      method: "patch",
      url: "/api/v1/category",
      data: {
        categoryId,
        categoryTitle,
      },
    })
      .then(({ data: categoryList }) => {
        dispatch({ type: RENAME_CATEGORY, categoryList });
      })
      .catch((err) => console.log(err));
  };
}

export function delCategory(categoryId) {
  return (dispatch) => {
    axios({
      method: "delete",
      url: "/api/v1/category",
      data: {
        categoryId,
      },
    }).then(({ data: categoryList }) => {
      const rootList = categoryList.filter(
        (el) => el.parentCategoryId === null
      );
      dispatch({ type: DEL_CATEGORY, categoryList });
      dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
    });
  };
}

export function updateChosenCategory(chosenCategory) {
  return { type: UPDATE_CHOSEN_CATEGORY, chosenCategory };
}

export function setIsFilterStatusDone(data) {
  return { type: SET_IS_FILTER_STATUS_DONE, data };
}

export function setIsCreateTaskModalOpen(data) {
  return { type: SET_IS_CREATE_TASK_MODAL, data };
}
