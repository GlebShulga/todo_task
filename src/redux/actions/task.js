import axios from "axios";
import {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_SEARCH_CRITERIA,
} from "../types/task";

export function fetchTaskList() {
  return (dispatch, getState) => {
    axios("/api/v1/task")
      .then(({ data }) => {
        const store = getState();
        const url = store.router.location.pathname.substring(1);
        const taskUrl = url?.match("(?<=/)(.*?)(?=/edit)");
        dispatch({ type: GET_TASK_LIST, taskList: data });
        if (url.includes("edit")) {
          const chosenTask = data.find((task) => task.title === taskUrl[0]);
          dispatch({ type: UPDATE_CHOSEN_TASK, chosenTask });
        }
      })
      .catch((err) => console.log(err));
  };
}

export function addTask(title, categoryId) {
  return (dispatch) => {
    axios({
      method: "post",
      url: "/api/v1/task",
      data: {
        title,
        categoryId,
      },
    })
      .then(({ data: taskList }) => {
        dispatch({ type: ADD_TASK, taskList });
        dispatch({ type: GET_TASK_LIST, taskList });
      })
      .catch((err) => console.log(err));
  };
}

export function patchTask(taskId, status, title, description, categoryId) {
  return (dispatch) => {
    axios({
      method: "patch",
      url: "/api/v1/task",
      data: {
        taskId,
        status,
        title,
        description,
        categoryId,
      },
    })
      .then(({ data: taskList }) => {
        dispatch({ type: PATCH_TASK, taskList });
      })
      .catch((err) => console.log(err));
  };
}

export function updateChosenTask(chosenTask) {
  return { type: UPDATE_CHOSEN_TASK, chosenTask };
}

export function setEditingTaskMode(data) {
  return { type: SET_EDITING_TASK_MODE, data };
}

export function setSearchCriteria(searchCriteria) {
  return { type: SET_SEARCH_CRITERIA, searchCriteria };
}

export function setNewCategoryIdForTask(categoryId) {
  return { type: SET_NEW_CATEGORY_FOR_TASK, categoryId };
}
