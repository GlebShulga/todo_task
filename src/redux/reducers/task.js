import {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_SEARCH_CRITERIA,
} from "../types/task";

const initialState = {
  taskList: [],
  chosenTask: {},
  searchCriteria: "",
  isEditingTaskMode: false,
  newCategoryIdForTask: "",
};

// eslint-disable-next-line import/no-anonymous-default-export
export default (state = initialState, action) => {
  switch (action.type) {
    case GET_TASK_LIST:
    case ADD_TASK:
    case PATCH_TASK: {
      return { ...state, taskList: action.taskList };
    }
    case UPDATE_CHOSEN_TASK: {
      return { ...state, chosenTask: action.chosenTask };
    }
    case SET_EDITING_TASK_MODE: {
      return { ...state, isEditingTaskMode: action.data };
    }
    case SET_SEARCH_CRITERIA: {
      return { ...state, searchCriteria: action.searchCriteria };
    }
    case SET_NEW_CATEGORY_FOR_TASK: {
      return { ...state, newCategoryIdForTask: action.categoryId };
    }
    default:
      return state;
  }
};


