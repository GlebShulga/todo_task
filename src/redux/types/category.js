const ADD_CATEGORY = "ADD_CATEGORY";
const GET_CATEGORIES_LIST = "GET_CATEGORIES_LIST";
const GET_CATEGORIES_TITLE_LIST = "GET_CATEGORIES_TITLE_LIST";
const GET_ROOT_CATEGORY_LIST = "GET_ROOT_CATEGORY_LIST";
const RENAME_CATEGORY = "RENAME_CATEGORY";
const DEL_CATEGORY = "DEL_CATEGORY";
const UPDATE_CHOSEN_CATEGORY = "UPDATE_CHOSEN_CATEGORY";
const SET_IS_FILTER_STATUS_DONE = "SET_IS_FILTER_STATUS_DONE";
const SET_IS_CREATE_TASK_MODAL = "SET_IS_CREATE_TASK_MODAL";
const SET_IS_OPEN_TASK_TABLE = "SET_IS_OPEN_TASK_TABLE";

module.exports = {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
};
