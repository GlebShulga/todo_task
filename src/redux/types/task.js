const ADD_TASK = "ADD_TASK";
const GET_TASK_LIST = "GET_TASK_LIST";
const PATCH_TASK = "PATCH_TASK";
const UPDATE_CHOSEN_TASK = "UPDATE_CHOSEN_TASK";
const SET_EDITING_TASK_MODE = "SET_EDITING_TASK_MODE ";
const SET_NEW_CATEGORY_FOR_TASK = "SET_NEW_CATEGORY_FOR_TASK";
const SET_IS_SEARCH = "SET_IS_SEARCH";
const SET_SEARCH_CRITERIA = "SET_SEARCH_CRITERIA";

module.exports = {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_IS_SEARCH,
  SET_SEARCH_CRITERIA,
};
