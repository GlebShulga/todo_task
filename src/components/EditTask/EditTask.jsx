import React, { useEffect, useState } from "react";
import "./EditTask.scss";

const EditTask = ({
  task,
  patchTask,
  setEditingTaskMode,
  newCategoryIdForTask,
}) => {
  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [status, setStatus] = useState("new");
  const [lengthError, setLengthError] = useState(false);
  const taskId = task.taskId;
  const categoryId = newCategoryIdForTask;

  useEffect(() => {
    setTitle(task.title);
    setDescription(task.description)
    setStatus(task.status)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const isChecked = status === "done";

  const onChangeTitle = (e) => {
    setTitle(e.target.value);
    setLengthError(false);
  };

  const onChangeStatus = (e) => {
    e.target.checked ? setStatus("done") : setStatus("new");
  };

  const onChangeDescription = (e) => {
    setDescription(e.target.value);
  };

  const onClickEditTask = async () => {
    if (title?.length <= 20 && title?.length >= 3) {
      await patchTask(taskId, status, title, description, categoryId);
      setEditingTaskMode(false);
    } else {
      setLengthError(true);
    }
  };

  const taskDescription =
    task.description === ""
      ? "Write description of your task"
      : task.description;

  return (
    <div className="EditTask">
      <div className="EditTask-buttons">
        <button
          type="button"
          className="CreateTask-form_button"
          onClick={onClickEditTask}
        >
          Save changes
        </button>
        <button
          type="button"
          className="EditTask-buttons__close"
          onClick={() => {
            setEditingTaskMode(false);
          }}
        >
          Cancel
        </button>
      </div>
      <div className="EditTask-title">
        <input
          className="EditTask-title__input"
          type="text"
          value={title ?? null}
          onChange={onChangeTitle}
        />
        {lengthError && (
          <div className="EditTask-Error">
            The task length must not be shorter than 3 characters and must not
            exceed 20 characters.
          </div>
        )}
      </div>
      <div className="EditTask-status">
        <input
          type="checkbox"
          id="status"
          name="status"
          checked={isChecked}
          value={true}
          onChange={onChangeStatus}
        />
        <label htmlFor="status">Done</label>
      </div>
      <textarea
        className="EditTask-description__input"
        type="text"
        value={description ?? null}
        onChange={onChangeDescription}
        placeholder={taskDescription}
      />
    </div>
  );
};

export default EditTask;
