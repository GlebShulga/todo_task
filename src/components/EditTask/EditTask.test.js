import React from "react";
import configureStore from "redux-mock-store";
import { BrowserRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import { render, fireEvent } from "@testing-library/react";
import renderer, { act } from "react-test-renderer";
import { shallow } from "enzyme";
import { toMatchDiffSnapshot } from "snapshot-diff";
import EditTask from "./EditTask";
import { mockData } from "../../../tests/mockStore";
import { setEditingTaskMode } from "../../redux/actions/task";
import { SET_EDITING_TASK_MODE } from "../../redux/types/task";

expect.extend({ toMatchDiffSnapshot });

const mockHistoryPush = jest.fn();

const middlewares = [thunk];
let mockStore;
const mockStoreConf = configureStore(middlewares);

jest.mock("../../redux/actions/task", () => ({
  ...jest.requireActual("../../redux/actions/task"),
  setEditingTaskMode: jest.fn(),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe("EditTask Component", () => {
  beforeEach(() => {
    mockStore = mockStoreConf(mockData);
  });

  it("renders EditTask without crashing", () => {
    shallow(
      <Router>
        <Provider store={mockStore}>
          <EditTask />
        </Provider>
      </Router>
    );
  });

  it("EditTask snapshot", () => {
    const component = renderer.create(
      <Router>
        <Provider store={mockStore}>
          <EditTask />
        </Provider>
      </Router>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    act(() => {
      component.root.findAllByType("button")[0].props.onClick();
    });
    const treeUpdate = component.toJSON();
    expect(tree).toMatchDiffSnapshot(treeUpdate);
  });

  it("should click on Cancel Button", () => {
    let mockFn;
    setEditingTaskMode.mockImplementationOnce((callback) => {
      mockFn = jest.fn((dispatch) => {
        dispatch({
          type: SET_EDITING_TASK_MODE,
          isEditingTaskMode: false,
        });
      });
      return mockFn;
    });

    const { getByTestId } = render(
      <Router>
        <Provider store={mockStore}>
          <EditTask />
        </Provider>
      </Router>
    );

    fireEvent.click(getByTestId("CancelChangesButton"));

    expect(mockStore.getActions()).toEqual([
      {
        type: SET_EDITING_TASK_MODE,
        isEditingTaskMode: false,
      },
    ]);
    expect(mockHistoryPush).toHaveBeenCalledTimes(1);
  });
});
