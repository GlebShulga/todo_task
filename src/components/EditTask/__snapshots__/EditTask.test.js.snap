// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTask Component EditTask snapshot 1`] = `
<div
  className="edit_task"
>
  <div
    className="edit_task-buttons"
  >
    <button
      className="form_button_add"
      onClick={[Function]}
      type="button"
    >
      Save changes
    </button>
    <button
      className="edit_task-buttons__close"
      data-testid="CancelChangesButton"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
  <div
    className="edit_task-title"
  >
    <input
      className="edit_task-title__input"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="status"
  >
    <input
      checked={false}
      id="status"
      name="status"
      onChange={[Function]}
      type="checkbox"
      value={true}
    />
    <label
      htmlFor="status"
    >
      Done
    </label>
  </div>
  <textarea
    className="edit_task-description__input"
    onChange={[Function]}
    placeholder="Write description of your task"
    type="text"
  />
</div>
`;

exports[`EditTask Component EditTask snapshot 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -27,10 +27,15 @@
        className=\\"edit_task-title__input\\"
        onChange={[Function onChangeTitle]}
        type=\\"text\\"
        value=\\"\\"
      />
+     <div
+       className=\\"edit_task-error\\"
+     >
+       The task length must not be shorter than 3 characters and must not exceed 20 characters.
+     </div>
    </div>
    <div
      className=\\"status\\"
    >
      <input"
`;
